wordpress-mobile/WordPress-FluxC-Android
# Woo/feature shipping labels

This is the feature branch of the shipping label endpoints. This includes:
- endpoints to fetch shipping labels for an order.
- endpoint to refund a shipping label.
- endpoint to preview a shipping label.

~**This PR is in draft since this [refund PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1582) and this [preview PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1584) need to be merged.**~

#### Screenshots
<img width="300" src="https://user-images.githubusercontent.com/22608780/81919508-922a6700-95f5-11ea-87ad-7e9b1e18ae5f.png"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/81919516-93f42a80-95f5-11ea-8d81-7c362f026af6.png">. <img src="https://user-images.githubusercontent.com/22608780/82300375-6253d880-99d4-11ea-8853-8da0d70bed47.gif" width="300"/>. <img src="https://user-images.githubusercontent.com/22608780/82407201-33963a80-9a86-11ea-8739-8d355ee1af93.gif" width="300" />
 
#### Testing
- Since this PR adds a migration script, it would be nice to test updating the app.
- Run `WCShippingLabelStoreTest` and `WCShippingLabelSqlUtilsTest`.
- To test fetching shipping labels:
  - In the example app, click on `Woo` -> `Shipping Labels` -> `Select Site` -> `Fetch Shipping Labels`.
    - Enter an invalid order number and verify that an error message is displayed. 
    - Enter a valid order number without any shipping labels and verify that the list is empty, while the request is successful. 
    - Enter a valid order number with shipping labels and verify that the shipping labels are fetched successfully. 
    - Try deleting or refunding a shipping label for the order and verify it is updated in the app.
    - Try fetching shipping labels for multiple orders and verify the correct list is fetched and displayed.
- To test refund shipping label:
  - In the example app, click on `Woo` -> `Shipping Labels` -> `Select Site` -> `Refund Shipping Label`.
    - Enter a valid order number with shipping label already purchased.
    - Enter the label id of the shipping label purchased.
    - Verify that the refund is successful.
    - Now click on `Fetch Shipping Labels`.
    - Use Stethos to verify that the `WCShippingLabelModel` consists of a `refund` column that is not empty.  
- To test preview of a shipping label:
  - In the example app, click on `Woo` -> `Shipping Labels` -> `Select Site` -> `Print Shipping Label`.
    - Enter the label id of the shipping label purchased.
    - Verify that the request is successful. Notice that you are redirected to the preview pdf of the shipping label.
